### PR TITLE
Implement project status badge #7618

### PR DESCRIPTION
**Pull request by sergiogcx**
## Associated issues
Closes https://github.com/cityofaustin/atd-data-tech/issues/7618
Closes https://github.com/cityofaustin/atd-data-tech/issues/7640
Closes https://github.com/cityofaustin/atd-data-tech/issues/7639

## Testing
https://moped-test.austinmobility.io/moped

**Steps to test:**
- Create a new project, verify that it gets a "Potential" badge  (from #7635)
- Update project phase, verify that status badge updates
- Cancel project, verify that it no longer shows in project listing #7640

---
#### Ship list
- [ ] Code reviewed 
- [x] Product manager approved
- [x] Added to [QA test script](https://docs.google.com/spreadsheets/d/1n_O6MLh9cwwPf57HUM394Ea-z9uuoEV1-QW4axNZXLE/edit#gid=776973707)

## Preview(s):

https://user-images.githubusercontent.com/5282430/143905318-12ae4df2-0bdf-435f-98c0-0582bd77fa2e.mp4

https://user-images.githubusercontent.com/5282430/143905335-1fac59b1-5044-42b1-a57d-4ab0a7a5aab2.mp4

https://user-images.githubusercontent.com/5282430/143905347-76d853b8-9f94-4322-a483-b494466dec59.mp4

https://user-images.githubusercontent.com/5282430/143905359-015c8fe6-060f-41b7-91ce-0fd4c76ded7c.mp4

### CODE REVIEW

**Comment by sergiogcx** [comment]
@amenity @zappfyllc @johnclary 

Same thing here as previously, this is the first stable rendering of this feature. Looking for product feedback.

File: moped-editor/src/views/projects/projectView/ProjectStatusBadge.js
Comment (johnclary) [inline moped-editor/src/views/projects/projectView/ProjectStatusBadge.js:113]: looks like a dupe entry for on-hold?

File: moped-editor/src/views/projects/projectView/ProjectStatusBadge.js
Comment (johnclary) [inline moped-editor/src/views/projects/projectView/ProjectStatusBadge.js:176]: perhaps we should extract these magic numbers to a settings file so they're easier to track down if we make a breaking change down the road?

File: moped-editor/src/views/projects/projectView/ProjectStatusBadge.js
Comment (johnclary) [inline moped-editor/src/views/projects/projectView/ProjectStatusBadge.js:225]: ah, yes—ditto here: let's put these status ID properties elsewhere

File: moped-editor/src/views/projects/projectView/ProjectView.js
Comment (johnclary) [inline moped-editor/src/views/projects/projectView/ProjectView.js:428]: @sergiogcx can you conditionally hide the "Place on hold" item if the project is already on hold?

It's not critical, but i think you have all the state here to easily do this.

**Comment by johnclary** [comment]
@sergiogcx sorry i just realized you were looking for initial product feedback. not code feedback. 

this is super cool and you've handled this complex business logic really well. i just tried quite hard to break it, and i could not. it looks great to me. i do have a product question, but i'll drop it in https://github.com/cityofaustin/atd-data-tech/issues/7618 because i don't want to spam this PR.

File: moped-editor/src/views/projects/projectView/ProjectStatusBadge.js
Comment (sergiogcx) [reply moped-editor/src/views/projects/projectView/ProjectStatusBadge.js:176]: Ditto. Didn't want to optimize too soon, just wanted to make sure I delivered the feature correctly first.

**Comment by amenity** [comment]
Looking good!

Can we change the behavior on "Cancel"? I'd like to stay on the project page and see the "Canceled badge"

**Comment by amenity** [comment]
Screen Shot 2021-11-29 at 2 44 30 PM
Should this typo on the "Projects" page — `Onhold` instead of `On hold` — be a separate bug report?

**Comment by sergiogcx** [comment]
> Screen Shot 2021-11-29 at 2 44 30 PM
> 
> Should this typo on the "Projects" page — `Onhold` instead of `On hold` — be a separate bug report?

@amenity it should be a separate issue, I did not touch that part of the code at all and I don't know where the scope is gonna go if I try to solve it.

**Comment by sergiogcx** [comment]
I also noticed the icons remain the same, trying to solve it rn.

**Comment by amenity** [comment]
@sergiogcx - Let me make a whole separate issue for that page.

File: moped-editor/src/views/projects/projectView/ProjectView.js
Comment (amenity) [reply moped-editor/src/views/projects/projectView/ProjectView.js]: @sergiogcx @johnclary - so how does one take a project off hold? 🤔

Comment (amenity) [reply moped-editor/src/views/projects/projectView/ProjectView.js]: I guess setting a phase to "Active" works... I would expect the menu item to change to "Take off hold".

**Comment by amenity** [comment]
Currently, if I delete the active phase, even if there are other phases in the project, the status badge shows as "On hold". @johnclary @zappfyllc - I think we have two options:
- Hide the badge if the project status is active but there is no phase set to current
- Prevent user from turning off "current" on all phases.

**Review by amenity — CHANGES_REQUESTED** [review]
Fantastic, @sergiogcx... I'm good to save anything further for follow-up issues. My only request for this PR would be to remove "Current phase" and "Status" from the "Summary" page.
Screen Shot 2021-11-30 at 2 25 20 AM
.

**Comment by zappfyllc** [comment]
> Currently, if I delete the active phase, even if there are other phases in the project, the status badge shows as "On hold". @johnclary @zappfyllc - two options come to mind:

@amenity Not trying to shut down the options you provided here, but from my perspective, you identified a bug (or maybe unintended poor requirements on my part). If a newly created project has no "Current" phase, it should show its status, which should undoubtedly be "Active". For some reason its status here must be set to the wrong value.

Unless a project's status is "On hold", the project should never display as "On hold". Phases being removed or modified shouldn't have any impact on that front. @sergiogcx can you confirm what's making this display? To replicate seeing the "On hold" label:

1. Create a new project
2. Toggle the "Potential" phase (so that it is NOT current)

**Comment by amenity** [comment]
Aha — displaying "Active" with a green badge never occurred to me. 🙃  It feels kiiiiinda messy for some reason, but maybe not any less so than the other low-ish level of effort fixes I can come up with. 

From a data perspective, IMHO, all active projects should have a current phase identified. So maybe I'm just tiptoeing around weird ways of enforcing/enticing that.

**Comment by zappfyllc** [comment]
> Aha — displaying "Active" with a green badge never occurred to me. 🙃 It feels kiiiiinda messy for some reason, but maybe not any less so than the other low-ish level of effort fixes I can come up with.
> 
> From a data perspective, IMHO, all active projects should have a current phase identified. So maybe I'm just tiptoeing around weird ways of enforcing/enticing that.

I 100% agree re: enforcing that all active projects having a current phase identified, but I think that's more of our issue with the Timeline tab and the "Current phase" toggle that we hate :laughing:. We effectively gave them the ability to toggle their way out of giving us more specific information about where exactly a project is. But I think that's ok to punt on for this issue by displaying "Active" when the "Current phase" query returns nil. If we find an intuitive way to enforce identifying a current phase in the future, we'll never have to encounter the label displaying "Active" with a green badge anyway.

**Comment by sergiogcx** [comment]
@amenity @zappfyllc The on-hold behavior is intentional, as per [this part of the code](https://github.com/cityofaustin/atd-moped/pull/475/files#diff-42256b926efe0131d06eacb259af2b962b743165a68c9607c19c0e59ae6dbc00R588-R601). It is there because there was nothing in the issue that indicates what the behavior should be.

While the issues were not fully scoped, I wrote it this way so I could present it to you and have this conversation.

**Comment by sergiogcx** [comment]
When the current phase is removed, the code has to decide what to put in its stead. It could be empty and show no status, or it could pick one of the other statuses (if any), and if so: which one?

**Comment by amenity** [comment]
@sergiogcx - I believe @zappfyllc is suggesting that instead of displaying a phase, we display "Active".

**Comment by zappfyllc** [comment]
> @amenity @zappfyllc The on-hold behavior is intentional, as per this part of the code. It is there because there was nothing in the issue that indicates what the behavior should be. In fact, it doesn't even touch the on status change cases (on update, on delete, etc.). While the issues were not fully scoped, I wrote it this way so I could present it to you and have this conversation.

@sergiogcx This is my bad, but https://github.com/cityofaustin/atd-data-tech/issues/7635 actually was meant to addressed this and I accidentally struck out one of the items in a hurry because of the discussion around `moped_status` and a separate `moped_project_status` table. I think I was under the impression that newly created projects are being set as "Active" anyway and that we would default to the status name if a current phase wasn't found to display, but to your point, I did miss explicitly writing that acceptance criteria. Will try to do better in the future! Good idea on prompting this conversation :).

Nevertheless, the default status for any newly created projects should be "Active" and that should display as a badge in green (with the play icon) where there is no current phase identified. Sorry for missing that acceptance criteria, but that should be the default behavior, not "On hold".

**Comment by amenity** [comment]
> Nevertheless, the default status for any newly created projects should be "Active" and that should display as a badge in green (with the play icon) where there is no current phase identified. Sorry for missing that acceptance criteria, but that should be the default behavior, not "On hold".

Hooooold up... I thought that newly created projects get a "Potential" phase and that's set as the badge? This is working as expected as Sergio has it.

**Comment by sergiogcx** [comment]
Settling with this for now, as per our [conversation in slack](https://austininnovation.slack.com/archives/CNUEPKLB1/p1638291830022300). @amenity @zappfyllc

**Comment by amenity** [comment]
@sergiogcx @zappfyllc - thanks for all the hard work on this. I think it's super. 

Here are some follow-up issues, in order of priority
- [Render project status badge on "Projects" page](https://github.com/cityofaustin/atd-data-tech/issues/7827)
- [Update confirmation "Delete" confirmation dialog](https://github.com/cityofaustin/atd-data-tech/issues/7837)
- [Replace "Current status" filter in advanced search with project badge values](https://github.com/cityofaustin/atd-data-tech/issues/7831)
- [Enable users to see canceled projects](https://github.com/cityofaustin/atd-data-tech/issues/7832)

**Review by amenity — APPROVED** [review]

**Review by zappfyllc — APPROVED** [review]

File: moped-editor/src/views/projects/projectView/ProjectStatusBadge.js
Comment (chiaberry) [inline moped-editor/src/views/projects/projectView/ProjectStatusBadge.js:199]: Why is it beneficial to use `!!!` over `!`?

File: moped-editor/src/views/projects/projectView/ProjectView.js
Comment (chiaberry) [inline moped-editor/src/views/projects/projectView/ProjectView.js:260]: should this be "deleted project" ? I can update in my branch, just wanted to confirm it was a typo and not intended phrasing

**Comment by mateoclarke** [comment]
Hey folks, I think this PR was merged without a dev review like is our normal policy. I see Sergio assigned us to review at 9:45am and merged at 1:13pm which isn't really enough time. Just wanted to make sure we give time for feedback and review. Maybe I'm speaking without enough info if there was a verbal discussion/approval instead of a code review.

**Comment by sergiogcx** [comment]
@mateoclarke isn't @zappfyllc a developer?

> Hey folks, I think this PR was merged without a dev review like is our normal policy. I see Sergio assigned us to review at 9:45am and merged at 1:13pm which isn't really enough time. Just wanted to make sure we give time for feedback and review. Maybe I'm speaking without enough info if there was a verbal discussion/approval instead of a code review.

@mateoclarke is JD not a dev?

File: moped-editor/src/views/projects/projectView/ProjectStatusBadge.js
Comment (sergiogcx) [reply moped-editor/src/views/projects/projectView/ProjectStatusBadge.js:199]: It's a little too fancy and maybe unnecessary, it's an inverted truthy operation... or an ugly catch-all method. We are not assuming status is a string, null, empty array, etc.

https://developer.mozilla.org/en-US/docs/Glossary/Truthy

File: moped-editor/src/views/projects/projectView/ProjectView.js
Comment (sergiogcx) [reply moped-editor/src/views/projects/projectView/ProjectView.js:260]: Sweet, that would be amazing. Thanks @chiaberry

**Comment by mateoclarke** [comment]
Ah, I understand now. Yes, JD is a developer by trade. By all means, I didn’t intend to overlook his skills. On Moped specifically, JD’s role has been product management and DB schema development (among other things). Especially for Front-End and React heavy features, I think we should allow for time from the broader dev team to have a chance to review. In looking back over the reviews, It also seems like JD's feedback was specific to product features and not related to the code. Again, happy to be corrected if JD did do a review of the code in addition to the feature functionality.

File: moped-editor/src/views/projects/projectView/ProjectStatusBadge.js
Comment (chiaberry) [reply moped-editor/src/views/projects/projectView/ProjectStatusBadge.js:199]: thanks, I was just under the impression that !!! was functionally similar to ! and was wondering if there was a different behavior from one to the other.

**Comment by zappfyllc** [comment]
@mateoclarke no apology needed whatsoever, and I'm sorry for the confusion @sergiogcx, because I did approve this PR. Mateo is absolutely correct and I would not want to be the sole dev approval since I'm not taking a thorough enough pass at the code (though I do always read it and have commented on it in the past, so I can see where that could easily be confused). I was definitely approving from the product side: I don't have enough React chops to be commenting extensively on your implementations or approving them. Sorry for the confusion y'all. :)

File: moped-editor/src/views/projects/projectView/ProjectStatusBadge.js
Comment (mateoclarke) [reply moped-editor/src/views/projects/projectView/ProjectStatusBadge.js:199]: Yeah this is really small stylistic thing, but there is no difference between `!!!` and `!` in the way truthiness gets evaluated. My opinion is that it might be confusing, but again, it doesn't really matter.
TIL: https://stackoverflow.com/questions/21154510/the-use-of-the-triple-exclamation-mark

**Comment by amenity** [comment]
FWIW, I've been checking the lil' "Code reviewed" checkbox I added to the PR template if I see a dev has reviewed. The idea was that we'd want all boxes checked/ruled-out before merging. Not helpful to me if it's not helpful to y'all, just throwing it out there!

File: moped-editor/src/views/projects/projectView/ProjectStatusBadge.js
Comment (sergiogcx) [reply moped-editor/src/views/projects/projectView/ProjectStatusBadge.js:199]: This is a zen moment! 🙇 I will have it changed in a new PR, I think there are other instances of this in the code. CDN avatar iirc. Thanks chia, mateo!

File: moped-editor/src/views/projects/projectView/ProjectView.js
Comment (amenity) [reply moped-editor/src/views/projects/projectView/ProjectView.js:260]: There's actually an issue for this text that @tillyw has: https://github.com/cityofaustin/atd-data-tech/issues/7837

File: moped-editor/src/views/projects/projectView/ProjectStatusBadge.js
Comment (chiaberry) [inline moped-editor/src/views/projects/projectView/ProjectStatusBadge.js:205]: Why does theme get passed as `null` here? 

the function definition for getStyle says if theme is passed as null, the function returns null, so shouldn't this always coalesce to defaultIcon?

Comment (sergiogcx) [reply moped-editor/src/views/projects/projectView/ProjectStatusBadge.js:205]: @chiaberry Ah, excellent call. I need to change the docs. That was in fact the behavior until I corrected it. There used to be a statement like this `if(!status) return null;` and that always gave me the default icon no matter what.

Comment (sergiogcx) [reply moped-editor/src/views/projects/projectView/ProjectStatusBadge.js:205]: That function is the most cringe-worthy of the PR, I could not think of a better way to make the component reusable and easy to configure. I'd greatly appreciate any suggestions. 🙏

Comment (johnclary) [reply moped-editor/src/views/projects/projectView/ProjectStatusBadge.js:205]: i think it's fine as is, but since you asked...this might be slightly better:

```javascript
const componentMapName = getComponentMapName = (status, phase, projectStatuses);

const statusProperties = {
status: status,
phase: componentMapName, // It's here so it can be corrected by status id
};

const ChipIcon =
    getStyle(null, componentMapName)?.icon ?? defaultIcon;

```

**Pull request merged**
